riptide-backup/src/main/java/org/zalando/riptide/backup/BackupRequestPlugin.java

```diff
@@ -0,0 +1,83 @@
+package org.zalando.riptide.backup;
+
+import com.google.gag.annotation.remark.Facepalm;
+import org.springframework.http.client.ClientHttpResponse;
+import org.zalando.fauxpas.ThrowingRunnable;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestArguments;
+import org.zalando.riptide.RequestExecution;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+
+import static java.util.Objects.nonNull;
+
+public final class BackupRequestPlugin implements Plugin {
+
+    private final ScheduledExecutorService scheduler;
+    private final long delay;
+    private final TimeUnit unit;
+
+    public BackupRequestPlugin(final ScheduledExecutorService scheduler, final long delay, final TimeUnit unit) {
+        this.scheduler = scheduler;
+        this.delay = delay;
+        this.unit = unit;
+    }
+
+    @Override
+    public RequestExecution prepare(final RequestArguments arguments, final RequestExecution execution) {
+        switch (arguments.getMethod()) {
+            case GET:
+            case HEAD:
+                return withBackup(execution);
+            default:
+                return execution;
+        }
+    }
+
+    private RequestExecution withBackup(final RequestExecution execution) {
+        return () -> {
+            final CompletableFuture<ClientHttpResponse> original = execution.execute();
+            final CompletableFuture<ClientHttpResponse> backup = new CompletableFuture<>();
+            original.whenComplete(cancel(delay(backup(execution, backup))));
```

Comment:
Should the original be canceled in case the backup completes first?

Reply:
I'm not cancelling the backup request directly, but rather the schedule to send it. Cancelling an ongoing request will have virtually no impact on the server side, since it's unable to detect it until the response will be written back.

Reply:
Also, cancelling a `CompletableFuture` is basically useless, since it has no connection to the thread doing the actual work and therefore there is nothing to cancel.

Reply:
ok, got it. The actual request is async again and so currently won't be cancelled.

One more suggestion, introduce a variable for the result of `delay` to make it clearer it's not a pure function but schedules a thread as a side-effect. Suggested name `scheduledBackup`?